libpromises/evalfunction.c

```diff
@@ -1300,27 +1300,45 @@ static FnCallResult FnCallBundlesMatching(EvalContext *ctx, const Policy *policy
 
 /*********************************************************************/
 
-static FnCallResult FnCallPackagesMatching(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, const FnCall *fp, const Rlist *finalargs)
+static bool AddPackagesmatchingJsonLine(pcre *matcher, JsonElement *json, char *line)
 {
-    const bool installed_mode = (0 == strcmp(fp->name, "packagesmatching"));
-    pcre *matcher;
+    
+    if (strlen(line) > CF_BUFSIZE - 80)
     {
-        const char *regex_package = RlistScalarValue(finalargs);
-        const char *regex_version = RlistScalarValue(finalargs->next);
-        const char *regex_arch = RlistScalarValue(finalargs->next->next);
-        const char *regex_method = RlistScalarValue(finalargs->next->next->next);
-        char regex[CF_BUFSIZE];
+        Log(LOG_LEVEL_ERR,
+            "Line from package inventory is too long (%zd) to be sensible",
+            strlen(line));
+        return false;
+    }
+    
 
-        // Here we will truncate the regex if the parameters add up to over CF_BUFSIZE
-        snprintf(regex, sizeof(regex), "^%s,%s,%s,%s$",
-                 regex_package, regex_version, regex_arch, regex_method);
-        matcher = CompileRegex(regex);
-        if (matcher == NULL)
+    if (StringMatchFullWithPrecompiledRegex(matcher, line))
+    {
+        Seq *list = SeqParseCsvString(line);
+        if (SeqLength(list) != 4)
         {
-            return FnFailure();
+            Log(LOG_LEVEL_ERR,
+                "Line from package inventory '%s' did not yield correct number of elements.",
+                line);
+            SeqDestroy(list);
+            return true;
```

Comment:
Actually not. In the original implementation, this was returning ```continue``` in the for loop whereas other ```fail``` conditions here were returning ```break```. I wanted to leave original behavior unchanged.